polygon/sync/sync.go

```diff
@@ -159,29 +149,24 @@ func (s *Sync) onNewHeaderEvent(
 }
 
 func (s *Sync) Run(ctx context.Context) error {
-	oldTip, err := s.storage.TipBlockNumber(ctx)
+	tip, err := s.execution.CurrentHeader(ctx)
 	if err != nil {
 		return err
 	}
 
-	if err = s.downloader.DownloadUsingCheckpoints(ctx, oldTip); err != nil {
-		return err
-	}
-
-	newTip, err := s.storage.TipBlockNumber(ctx)
-	if err != nil {
-		return err
-	}
-	if err = s.downloader.DownloadUsingMilestones(ctx, newTip); err != nil {
+	if newTip, err := s.downloader.DownloadUsingCheckpoints(ctx, tip.Number.Uint64()); err != nil {
 		return err
+	} else if newTip != nil {
+		tip = newTip
 	}
 
-	if err = s.commitExecution(ctx, oldTip); err != nil {
+	if newTip, err := s.downloader.DownloadUsingMilestones(ctx, tip.Number.Uint64()); err != nil {
 		return err
+	} else if newTip != nil {
```

Comment:
btw I would suggest to change the signature for  `DownloadUsingMilestones` and `DownloadUsingCheckpoints` to a more "standard" one which is `*types.Header, bool, err` so that we can do 

```
newTip, ok, err := s.downloader.DownloadUsingMilestones(...)
if err != nil {
    return err
}
if !ok { 
   return nil
}

// ... process new tip ...
```